### PR TITLE
merge dev_synth_1_module (version 0.2.0) to main

### PR DESCRIPTION
**Pull request by herbberg**
added script synth_1_module.py (for re-synthesis of only one module)

### CODE REVIEW

File: ugt_fwtools/synth_1_module.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    args.vivado_base_dir = os.getenv("UGT_VIVADO_BASE_DIR")
+    if not args.vivado_base_dir:
+        raise RuntimeError("\033[1;31m Environment variable 'UGT_VIVADO_BASE_DIR' not set. Set with: 'export UGT_VIVADO_BASE_DIR=...' \033[0m")
+
+    # Vivado settings
+    args.settings64 = os.path.join(args.vivado_base_dir, args.vivado, "settings64.sh")
+    if not os.path.isfile(args.settings64):
+        raise RuntimeError(
+            f"\033[1;31m no such Xilinx Vivado settings file {args.settings64!r}\n \033[0m"
+            f"   \033[1;31m check if Xilinx Vivado {args.vivado} is installed on this machine. \033[0m"
```
Comment (arnobaer) [inline ugt_fwtools/synth_1_module.py:50]: Please try avoid using[ ANSI escape codes](https://en.wikipedia.org/wiki/ANSI_escape_code) in exception messages, better to separate concerns; exception messages should contain concise information about an error while ANSI codes can be used by some TTYs to format the output. As the output is not automatically a TTY (like a pipe to a text file, log file, other function handling an exception) applying ANSI escape codes should be applied only when testing `sys.stdout.isatty()` or by using a library for handling colored output like [coloredlogs](https://pypi.org/project/coloredlogs/).

File: ugt_fwtools/synth_1_module.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        )
+
+    if os.path.exists(module_path):
+        subprocess.run(["rm", "-rf", module_path]).check_returncode()
```
Comment (arnobaer) [inline ugt_fwtools/synth_1_module.py:54]: No need to invoke a subprocess
```
import shutil
...
if os.path.exists(module_path):
    shutil.rmtree(module_path)
```

File: ugt_fwtools/synth_1_module.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    logging.info("===========================================================================")
+    logging.info("running IPBB project, synthesis and implementation, creating bitfile for module %s ...", module_id)
+
+    implement_module(module_id, module_name, args)
+
+    logging.info("===========================================================================")
+    show_screen_sessions()
+
+    os.chdir(args.ipbb_dir)
+
```
Comment (arnobaer) [inline ugt_fwtools/synth_1_module.py:70]: This can be removed.

File: ugt_fwtools/synth_1_module.py (hunk 1 above)
Comment (arnobaer) [reply ugt_fwtools/synth_1_module.py:50]: A simple solution without dependencies could be...
```python
tty_bold_red = "\033[1;31m"
tty_reset = "\033[0m"

def print_error(message):
    if sys.stdout.isatty():
        message = f"{tty_bold_red}{message}{tty_clear}"
    print(message)
...
if error_condition:
    message = "my error description"
    print_error(message):  # will print in red on a TTY
    raise RuntimeError(message)  # can be handled as supposed
```

**Review by arnobaer — CHANGES_REQUESTED** [review]

File: ugt_fwtools/synth_1_module.py (hunk 1 above)
Comment (herbberg) [reply ugt_fwtools/synth_1_module.py]: done!

File: ugt_fwtools/synth_1_module.py (hunk 2 above)
Comment (herbberg) [reply ugt_fwtools/synth_1_module.py]: done!

File: ugt_fwtools/synth_1_module.py (hunk 3 above)
Comment (herbberg) [reply ugt_fwtools/synth_1_module.py]: done!

**Pull request merged**